src/util/system.h

```diff
@@ -102,6 +110,32 @@ std::string ShellEscape(const std::string& arg);
 #if HAVE_SYSTEM
 void runCommand(const std::string& strCommand);
 #endif
+
+#if defined(BOOST_POSIX_API) && defined(FD_CLOEXEC)
+/**
+ * Ensure a boost::process::child has its non-std fds all closed when exec
+ * is called.
+ */
+struct bpe_close_excess_fds : boost::process::extend::handler
+{
+    template<typename Executor>
+    void on_exec_setup(Executor&exec) const
+    {
+        try {
+            for (auto it : fs::directory_iterator("/dev/fd")) {
+                int64_t fd;
+                if (!ParseInt64(it.path().filename().native(), &fd)) continue;
+                if (fd <= 2) continue;  // leave std{in,out,err} alone
+                ::fcntl(fd, F_SETFD, ::fcntl(fd, F_GETFD) | FD_CLOEXEC);
```

Comment:
As I understand this code is executed in the child process. What's the advantage of using `fcntl(…FD_CLOEXEC` instead of straight-up closing the fd here?

Reply:
> Instead of closing directly, we have to set the `FD_CLOEXEC` flag and let `exec` close them; otherwise, boost::process's internal pipe would be closed and we get the wrong exception thrown (not actually a problem outside of our tests, but might as well do this cleanly).